File: 100. Same Tree/100. Same Tree.py
from typing import Optional, List
from collections import deque  

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        queue_p = deque([p])  
        queue_q = deque([q])  
        while queue_p and queue_q:
            node_p = queue_p.popleft()
            node_q = queue_q.popleft()
            if not node_p and not node_q:
                continue
            if not node_p or not node_q or node_q.val!=node_p.val:
                return False
            
            queue_p.append(node_p.left)
            queue_q.append(node_q.left)
            queue_p.append(node_p.right)
            queue_q.append(node_q.right)

        return not queue_p and not queue_q

def build_tree(values: List[Optional[int]]) -> Optional[TreeNode]:
    if not values:
        return None

    root = TreeNode(values[0])
    queue = deque([root])
    i = 1

    while i < len(values):
        node = queue.popleft()
        if values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1

        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1

    return root

# Test Cases
p1 = build_tree([1,2,3])
q1 = build_tree([1,2,3])

p2 = build_tree([1,2])
q2 = build_tree([1,None,2])

p3 = build_tree([1,2,1])
q3 = build_tree([1,1,2])

solution = Solution()
print(solution.isSameTree(p1, q1))  # True
print(solution.isSameTree(p2, q2))  # False
print(solution.isSameTree(p3, q3)) 

File: 100. Same Tree/100. Same Tree.js
// Definition for a binary tree node.
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val);
    this.left = (left === undefined ? null : left);
    this.right = (right === undefined ? null : right);
}

/**
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {boolean}
 */
var isSameTree = function(p, q) {
    let queueP = [p]
    let queueQ = [q]

    while (queueP.length && queueQ.length) {
        let nodeP = queueP.shift()
        let nodeQ = queueQ.shift()

        if (!nodeP && !nodeQ) continue
        if (!nodeP || !nodeQ || nodeP.val !== nodeQ.val) return false

        queueP.push(nodeP.left)
        queueQ.push(nodeQ.left)
        queueP.push(nodeP.right)
        queueQ.push(nodeQ.right)
    }

    return queueP.length === 0 && queueQ.length === 0
};

function buildTree(values) {
    if (!values.length) return null;

    let root = new TreeNode(values[0]);
    let queue = [root];
    let i = 1;

    while (i < values.length) {
        let node = queue.shift();
        if (values[i] !== null) {
            node.left = new TreeNode(values[i]);
            queue.push(node.left);
        }
        i++;

        if (i < values.length && values[i] !== null) {
            node.right = new TreeNode(values[i]);
            queue.push(node.right);
        }
        i++;
    }

    return root;
}

// Test Cases
let p1 = buildTree([1, 2, 3]);
let q1 = buildTree([1, 2, 3]);
console.log(isSameTree(p1, q1));  // true

let p2 = buildTree([1, 2]);
let q2 = buildTree([1, null, 2]);
console.log(isSameTree(p2, q2));  // false

let p3 = buildTree([1, 2, 1]);
let q3 = buildTree([1, 1, 2]);
console.log(isSameTree(p3, q3));  // false
